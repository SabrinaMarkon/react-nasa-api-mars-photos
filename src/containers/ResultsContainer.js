import React, { Component, Fragment } from 'react';
import ResultHeadings from '../components/ResultHeadings';

export default class ResultsContainer extends Component {

  render() {
    let searchResults = Array.from(this.props.searchResults);

    // format the year property to only show 4 digit year.
    searchResults.forEach(result => {
      if (result.year) {
        result.year = result.year.slice(0, 4);
      }
    });

    let results = searchResults.map((row,i) => {
      return (
        <Fragment key={i}>
          <div className="grid-item" key={i}>{row.name}</div>
          <div className="grid-item">{i}</div>
          <div className="grid-item">{row.nametype}</div>
          <div className="grid-item">{row.recclass}</div>
          <div className="grid-item">{row.mass}</div>
          <div className="grid-item">{row.fall}</div>
          <div className="grid-item">{row.year}</div>
          <div className="grid-item">{row.reclat}</div>
          <div className="grid-item">{row.reclong}</div>
        </Fragment>
      );
    }

    );
    return (
    <section className="grid center-align p-2">
      <ResultHeadings />
      {results}
    </section>
    );
  }
}