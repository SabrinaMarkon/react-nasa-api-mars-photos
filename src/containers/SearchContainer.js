import React, { Component } from 'react';
import Search from '../components/Search';

export default class SearchContainer extends Component {

  constructor(props) {
    super(props);
    this.doSearch = this.doSearch.bind(this);
  }

  doSearch = (searchfield, searchinput) => {
    this.props.doSearch(searchfield, searchinput);
  }

  render() {
    return (
      <div className="center-align pt-2">
        <Search doSearch={this.doSearch} />
      </div>
    );
  }
}