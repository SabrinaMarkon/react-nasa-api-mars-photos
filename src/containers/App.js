import React, {Component} from 'react';
import Header from '../components/Header';
import Nav from '../components/Nav';
import UserMessage from '../components/UserMessage';
import SearchContainer from '../containers/SearchContainer';
import ResultsContainer from '../containers/ResultsContainer';
import Pagination from '../components/Pagination';
import Footer from '../components/Footer';
import axios from 'axios';

const PAGE_LIMIT = 100;

export default class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      searchResults: [],
      errorMessage: '',
      currentPage: 0,
      totalPages: 0
    }
    this.doSearch = this.doSearch.bind(this);
    this.onPageChanged = this.onPageChanged.bind(this);
  }

  // ?$select=count(*)

  componentDidMount() {
    this._isMounted = true;
    let API_URL = 'https://data.nasa.gov/resource/gh4g-9sfh.json?$order=name&$limit=' + PAGE_LIMIT + '&$offset=0';
    axios.get(API_URL)
    .then(res => {
      /* Add a check in the .then() handler so this.setState is not called if the component has been unmounted:
      That is, how should react 'react' when you call setState on a component that has already unmounted. The right way 
      to handle it would be to cancel the data fetching request if the component will be unmounted for some reason
      (like user navigating away) */
      if (this._isMounted) {
        const searchResults = res.data;
        if (searchResults && searchResults.length) {
          this.setState({
            searchResults,
            errorMessage: ''
          });
        } else {
          this.setState({
            searchResults: [],
            errorMessage: 'No Results Found'
          });
        }
      }
    })
    .catch(err => {
      this.setState({
        searchResults: [],
        errorMessage: err.response.data.code,
        currentPage: 0,
        totalPages: 0
      });
    })
  }
  
  componentWillUnmount() {
    this._isMounted = false;
  }

  doSearch = (searchField, searchInput) => {
    let API_URL = 'https://data.nasa.gov/resource/gh4g-9sfh.json?$order=name&$limit=' + PAGE_LIMIT + '&$offset=' + this.state.currentPage;
    if (searchInput) {
      // check for special characters.
      let originalSearchInput = searchInput;
      searchInput = originalSearchInput.replace(/[^a-z0-9,-. ]/gi, '');
      if (searchInput !== originalSearchInput) {
        // user included weird characters the server doesn't accept.
        this.setState({
          searchResults: [],
          errorMessage: 'Special characters are not allowed except (space, comma, decimal, dash, apostrophe).'
        }); 
        return; 
      }
      API_URL = 'https://data.nasa.gov/resource/gh4g-9sfh.json?$order=name&$limit=' + PAGE_LIMIT + '&$offset=' + this.state.currentPage + '&$where=upper(' + searchField + ')=upper(\'' + searchInput + '\')';
    }
    axios.get(API_URL)
    .then(res => {
      const searchResults = res.data;
      if (searchResults && searchResults.length) {
        this.setState({
          searchResults,
          errorMessage: ''
        });
      } else {
        this.setState({
          searchResults: [],
          errorMessage: 'No Results Found'
        });
      }
    })
    .catch(err => { 
      this.setState({
        searchResults: [],
        errorMessage: err.response.data.code,
        currentPage: 0,
        totalPages: 0
      });
      // console.error("Error response:");
      // console.error(err.response.data);
      // console.error(err.response.status);
      // console.error(err.response.headers);
    });
  }

  /* Called with data of the current pagination state only when the current page changes. */
  onPageChanged = data => {
    let {searchResults} = this.state;
    const {currentPage, totalPages, pageLimit} = data;
    /* -1 to make it zero based */
    const offset = (currentPage - 1) * pageLimit; 
    searchResults = searchResults.slice(offset, offset + pageLimit);
    this.setState({
      currentPage,
      searchResults,
      totalPages
    });
  }

  render() {
    return (
      <>
        <Nav />
        <Header />
        <div className="fixed-bg">
          {this.state.errorMessage 
            ? <>
              <UserMessage userMessage={this.state.errorMessage} />
              <SearchContainer doSearch={this.doSearch} />
              </>
            : <>
              <SearchContainer doSearch={this.doSearch} />
              <ResultsContainer searchResults={this.state.searchResults} />
              </>
          }
          <Pagination totalRecords={1000} pageLimit={PAGE_LIMIT} pageNeighbours={2} onPageChanged={this.onPageChanged} />
          <Footer />          
        </div>
      </>
    )
  }
}

