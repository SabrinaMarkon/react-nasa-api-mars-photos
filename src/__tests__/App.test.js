import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

// components
import App from '../containers/App';
import Nav from '../components/Nav';
import SearchContainer from '../containers/SearchContainer';
import ResultsContainer from '../containers/ResultsContainer';

function setup() {
  const wrapper = shallow(<App />);
  //return { wrapper, props }
  return wrapper;
}

describe('Testing main App container', () => {
  it('renders the main App component', () => {
    const fragment = document.createElement('Fragment');
    ReactDOM.render(<App />, fragment);
    ReactDOM.unmountComponentAtNode(fragment);
  });

  it('should render a child Header component', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should render a child Nav component', () => {
    const wrapper = shallow(<Nav />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should render a child SearchContainer component', () => {
    const wrapper = shallow(<SearchContainer />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should render a child ResultsContainer component', () => {
    const wrapper = shallow(<ResultsContainer />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should render a child Footer component', () => {
    const wrapper = shallow(<Footer />);
    expect(wrapper.exists()).toBe(true);
  });

  it('it should match the App snapshot', () => {
    const wrapper = setup();
    console.log(wrapper);
    expect(wrapper).toMatchSnapshot();
  });
});


