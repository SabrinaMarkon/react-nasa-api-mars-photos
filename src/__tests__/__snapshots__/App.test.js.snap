// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing main App container it should match the App snapshot 1`] = `
<Fragment>
  <Nav />
  <Header />
  <SearchContainer
    doSearch={[Function]}
  />
  <ResultsContainer />
  <Footer />
</Fragment>
`;
